### PR TITLE
Closes #6281: Task/lcp refactor rucss api cron queue

### LINKED ISSUE #6281: LCP & Above the Fold: Refactor the RUCSS API, Queue and CRON
**Is your feature request related to a problem? Please describe.**
Refactor the RUCSS API, Queue and CRON so that one request is done per page for all SaaS optimizations

### PR DESCRIPTION
**Pull request by jeawhanlee**
## Description

Refactor RUCSS, cron, queue and API module.

Fixes #6281 

## Type of change

sub-task

## Is the solution different from the one proposed during the grooming?

No

# Checklists

## Generic development checklist

- [x] My code follows the style guidelines of this project, with adapted comments and without new warnings.
- [ ] I have added unit and integration tests that prove my fix is effective or that my feature works.
- [ ] The CI passes locally with my changes (including unit tests, integration tests, linter).

## Test summary

- [ ] I triggered all changed lines of code at least once without new errors/warnings/notices.
- [ ] I validated all Acceptance Criteria of the related issues. (If applicable, provide proof).
- [ ] I validated all test plan the QA Review asked me to.

### CODE REVIEW

File: inc/Engine/Common/Database/Queries/AbstractQuery.php [hunk 1]
```diff
@@ -162,11 +187,6 @@ public function get_pending_jobs( int $count = 100 ) {
 			[
 				'number'         => ( $count - $inprogress_count ),
 				'status'         => 'pending',
-				'fields'         => [
-					'id',
-					'url',
-					'next_retry_time',
-				],
```
Comment (remyperona) [inline inc/Engine/Common/Database/Queries/AbstractQuery.php:169]: I don't think those should be removed

File: inc/Engine/Common/Database/Queries/AbstractQuery.php [hunk 2]
```diff
@@ -229,21 +259,24 @@ public function update_job_id( $id, $new_job_id ) {
 	/**
 	 * Change the status to be in-progress.
 	 *
-	 * @param int $id DB row ID.
-	 *
-	 * @return bool
+	 * @param string  $url Url from DB row.
+	 * @param boolean $is_mobile Is mobile from DB row.
+	 * @return int/boolean
```
Comment (remyperona) [inline inc/Engine/Common/Database/Queries/AbstractQuery.php]:
```suggestion
	 * @return bool|int
```

File: inc/Engine/Common/Database/Queries/AbstractQuery.php [hunk 3]
```diff
@@ -276,29 +309,38 @@ public function reset_job( int $id, string $job_id = '' ) {
...
-	 * @param int    $id DB row ID.
-	 * @param string $error_code error code.
-	 * @param string $error_message error message.
+	 * @param string  $url Url from DB row.
+	 * @param boolean $is_mobile Is mobile from DB row.
+	 * @param string  $error_code error code.
+	 * @param string  $error_message error message.
 	 *
-	 * @return bool
+	 * @return int/boolean
```
Comment (remyperona) [inline inc/Engine/Common/Database/Queries/AbstractQuery.php]:
```suggestion
	 * @return bool|int
```

File: inc/Engine/Common/JobManager/APIHandler/AbstractAPIClient.php [hunk 4]
```diff
@@ -1,17 +1,18 @@
...
 use WP_Rocket\Admin\Options_Data;
 use WP_Rocket\Logger\Logger;
+use WP_Rocket\Engine\Common\Context\ContextInterface;
 
 abstract class AbstractAPIClient {
 	/**
 	 * API URL.
 	 */
-	const API_URL = 'https://saas.wp-rocket.me/';
+	const API_URL = 'https://rucss-director-staging.public-default.live2-k8s-cph3.one.com/';
```
Comment (remyperona) [inline inc/Engine/Common/JobManager/APIHandler/AbstractAPIClient.php]: we will need to remember to update this later

File: inc/Engine/Common/JobManager/Context/RUCSSContext.php [hunk 5]
```diff
@@ -0,0 +1,23 @@
+<?php
+
+namespace WP_Rocket\Engine\Common\JobManager\Context;
```
Comment (remyperona) [inline inc/Engine/Common/JobManager/Context/RUCSSContext.php]: I think this should be in the RUCSS namespace, even if we use it here. We could use it later in the RUCSS classes too.

File: inc/Engine/Common/JobManager/Cron/Subscriber.php [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+	 * @since 3.9
+	 *
+	 * @return void
+	 */
+	public function schedule_clean_not_commonly_used_rows() {
+		if ( ! $this->rucss_context->is_allowed() && ! $this->atf_context->is_allowed() ) {
+			return;
+		}
+
+		if ( wp_next_scheduled( 'rocket_rucss_atf_clean_rows_time_event' ) ) {
```
Comment (remyperona) [inline inc/Engine/Common/JobManager/Cron/Subscriber.php]: I think we should make the SaaS related events a more generic name, like `rocket_sass_clean_rows_time_event`.

File: inc/Engine/Common/JobManager/Cron/Subscriber.php [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+	 * @param string  $optimization_type The type of optimization request to send.
+	 *
+	 * @return void
+	 */
+	public function check_job_status( string $url, bool $is_mobile, string $optimization_type ) {
+		$this->job_processor->check_job_status( $url, $is_mobile, $optimization_type );
+	}
+
+	/**
+	 * Adds new interval for RUCSS pending jobs cron
```
Comment (remyperona) [inline inc/Engine/Common/JobManager/Cron/Subscriber.php]:
```suggestion
	 * Adds new interval for SaaS pending jobs cron
```

File: inc/Engine/Common/JobManager/Cron/Subscriber.php [hunk 8]
```diff
@@ -0,0 +1,391 @@
...
+
+		if ( wp_next_scheduled( 'rocket_remove_rucss_atf_failed_jobs' ) ) {
+			return;
+		}
+
+		wp_schedule_event( time(), 'rocket_remove_rucss_atf_failed_jobs', 'rocket_remove_rucss_atf_failed_jobs' );
+	}
+
+	/**
+	 * Clear schedule of RUCSS CRONs on deactivation.
```
Comment (remyperona) [inline inc/Engine/Common/JobManager/Cron/Subscriber.php]:
```suggestion
	 * Clear schedule of SaaS CRONs on deactivation.
```

File: inc/Engine/Common/JobManager/Queue/Queue.php [hunk 9]
```diff
@@ -26,7 +26,7 @@ class Queue extends AbstractASQueue {
 	 *
 	 * @var string
 	 */
-	private $pending_job_cron = 'rocket_rucss_pending_jobs_cron';
+	private $pending_job_cron = 'rocket_rucss_atf_pending_jobs_cron';
```
Comment (remyperona) [inline inc/Engine/Common/JobManager/Queue/Queue.php]:
```suggestion
	private $pending_job_cron = 'rocket_saas_pending_jobs_cron';
```

File: inc/Engine/Common/JobManager/Queue/Queue.php [hunk 10]
```diff
@@ -60,15 +60,19 @@ public function schedule_pending_jobs_cron( int $interval ) {
...
+	 * @param string  $optimization_type The type of optimization request to send.
 	 *
-	 * @return string
+	 * @return int
 	 */
-	public function add_job_status_check_async( int $usedcss_row_id ) {
+	public function add_job_status_check_async( string $url, bool $is_mobile, string $optimization_type ) {
 		return $this->add_async(
-			'rocket_rucss_job_check_status',
+			'rocket_rucss_atf_job_check_status',
```
Comment (remyperona) [inline inc/Engine/Common/JobManager/Queue/Queue.php]:
```suggestion
			'rocket_saas_job_check_status',
```

File: inc/Engine/Common/JobManager/Context/RUCSSContext.php [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+
+use WP_Rocket\Engine\Common\Context\AbstractContext;
+
+class RUCSSContext extends AbstractContext {
```
Comment (CrochetFeve0251) [inline inc/Engine/Common/JobManager/Context/RUCSSContext.php]: @jeawhanlee This context shouldn't be inside the RUCSS part?

File: inc/Engine/Common/JobManager/Cron/Subscriber.php [hunk 12]
```diff
@@ -0,0 +1,391 @@
...
+	 *
+	 * @return array
+	 */
+	public static function get_subscribed_events(): array {
+		return [
+			'rocket_rucss_atf_pending_jobs'          => 'process_pending_jobs',
+			'rocket_rucss_atf_on_submit_jobs'        => 'process_on_submit_jobs',
+			'rocket_rucss_atf_job_check_status'      => [ 'check_job_status', 10, 3 ],
+			'rocket_rucss_job_check_status'          => 'rucss_check_job_status',
+			'rocket_rucss_atf_clean_rows_time_event' => 'cron_clean_rows',
```
Comment (CrochetFeve0251) [inline inc/Engine/Common/JobManager/Cron/Subscriber.php]: @jeawhanlee why theses hooks are related to rucss? Shouldn't they be abstract? (we are in common)

File: inc/Engine/Common/JobManager/Cron/Subscriber.php [hunk 13]
```diff
@@ -0,0 +1,391 @@
...
+	public function schedule_clean_not_commonly_used_rows() {
+		if ( ! $this->rucss_context->is_allowed() && ! $this->atf_context->is_allowed() ) {
+			return;
+		}
+
+		if ( wp_next_scheduled( 'rocket_rucss_atf_clean_rows_time_event' ) ) {
+			return;
+		}
+
+		wp_schedule_event( time(), 'weekly', 'rocket_rucss_atf_clean_rows_time_event' );
```
Comment (CrochetFeve0251) [inline inc/Engine/Common/JobManager/Cron/Subscriber.php]: @jeawhanlee same here

File: inc/Engine/Common/JobManager/Cron/Subscriber.php [hunk 14]
```diff
@@ -0,0 +1,391 @@
...
+		}
+
+		/**
+		 * Filters the cron interval.
+		 *
+		 * @since 3.11
+		 *
+		 * @param int $interval Interval in seconds.
+		 */
+		$interval = apply_filters( 'rocket_rucss_atf_pending_jobs_cron_interval', 1 * rocket_get_constant( 'MINUTE_IN_SECONDS', 60 ) );
```
Comment (CrochetFeve0251) [inline inc/Engine/Common/JobManager/Cron/Subscriber.php]: @jeawhanlee here too

File: inc/Engine/Common/JobManager/Cron/Subscriber.php [hunk 15]
```diff
@@ -0,0 +1,391 @@
...
+
+		$schedules['rocket_rucss_atf_pending_jobs'] = [
```
Comment (CrochetFeve0251) [inline inc/Engine/Common/JobManager/Cron/Subscriber.php]: @jeawhanlee here too

File: inc/Engine/Common/JobManager/Cron/Subscriber.php [hunk 16]
```diff
@@ -0,0 +1,391 @@
...
+		JobProcessor $job_processor,
+		Database $database,
+		ATFTable $atf_table,
+		ContextInterface $rucss_context,
+		ContextInterface $atf_context
+	) {
+		$this->job_processor = $job_processor;
+		$this->database      = $database;
+		$this->atf_table     = $atf_table;
+		$this->rucss_context = $rucss_context;
```
Comment (CrochetFeve0251) [inline inc/Engine/Common/JobManager/Cron/Subscriber.php]: @jeawhanlee why are we using the rucss context?

File: inc/Engine/Common/JobManager/Managers/AtfManager.php [hunk 17]
```diff
@@ -0,0 +1,94 @@
+<?php
+
+namespace WP_Rocket\Engine\Common\JobManager\Managers;
+
+use WP_Rocket\Logger\LoggerAware;
+use WP_Rocket\Logger\LoggerAwareInterface;
+use WP_Rocket\Engine\Media\AboveTheFold\Database\Queries\AboveTheFold as ATFQuery;
+use WP_Rocket\Engine\Common\Context\ContextInterface;
+
+class AtfManager extends AbstractManager implements ManagerInterface, LoggerAwareInterface {
```
Comment (CrochetFeve0251) [inline inc/Engine/Common/JobManager/Managers/AtfManager.php]: @jeawhanlee same with the Atf why it is inside the common namespace? Shouldn't be that logic inside the Above the fold namespace?

File: inc/Engine/Common/JobManager/Managers/RUCSSManager.php [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+namespace WP_Rocket\Engine\Common\JobManager\Managers;
+
+use WP_Rocket\Logger\LoggerAware;
+use WP_Rocket\Logger\LoggerAwareInterface;
+use WP_Rocket\Engine\Optimization\RUCSS\Database\Queries\UsedCSS as UsedCSS_Query;
+use WP_Rocket\Engine\Optimization\RUCSS\Controller\Filesystem;
+use WP_Rocket\Engine\Common\Context\ContextInterface;
+use WP_Rocket\Engine\Optimization\CSSTrait;
+
+class RUCSSManager extends AbstractManager implements ManagerInterface, LoggerAwareInterface {
```
Comment (CrochetFeve0251) [inline inc/Engine/Common/JobManager/Managers/RUCSSManager.php]: @jeawhanlee same here

File: inc/Engine/Common/JobManager/ServiceProvider.php [hunk 19]
```diff
@@ -0,0 +1,177 @@
...
+use WP_Rocket\Engine\Media\AboveTheFold\Context\Context;
+use WP_Rocket\Engine\Common\JobManager\Queue\Queue;
+use WP_Rocket\Engine\Common\JobManager\APIHandler\APIClient;
+use WP_Rocket\Engine\Optimization\RUCSS\Admin\Database;
+use WP_Rocket\Engine\Common\JobManager\Cron\Subscriber as CronSubscriber;
+use WP_Rocket\Engine\Media\AboveTheFold\Database\Tables\AboveTheFold as ATFTable;
+use WP_Rocket\Engine\Optimization\RUCSS\Database\Tables\UsedCSS as UsedCSSTable;
+
+
+class ServiceProvider extends AbstractServiceProvider {
```
Comment (CrochetFeve0251) [inline inc/Engine/Common/JobManager/ServiceProvider.php:18]: I think we need to make a clear split between what is inside the abstract logic and RUCSS and ATF ones.
Using factories here could be the way

**Review by CrochetFeve0251 — CHANGES_REQUESTED** [review]
i think we should make a clear split between what is abstract and concrete logic aka atf and rucss. 

The issue here is that we mixed namespaces but also concerns.
For me the queue should have no idea from what it processes thanks to abstraction so it will be only in charge of handling the queue. It will also prevent to have to change that logic if we need in the future to add a new set of actions to that queue.

Also you can insert families of classes using factories or using tags on the container which will prevent you from having to handle it inside the subscriber.

For the hooks you can use a variable like `"rocket{$this->my_prefix}_pending_jobs"` or find an abstract name so it won't be heavily linked to the RUCSS or ATF.

Also I don't see any tests is this something wanted? 

PS: It can seem like a lot of comments but as it is architectural things it impacts a bit everywhere. So I tried to keep only comments which illustrated each cases.

PS 2: I don't see any deprecated code when most hooks linked to RUCSS seems to have disappeared leading to potential breaks inside the configs from users. I think that would be beneficial for our support to make a bridge between the old hooks and the new ones.

File: inc/Engine/Common/JobManager/APIHandler/AbstractAPIClient.php (hunk 4 above)
Comment (CrochetFeve0251) [inline inc/Engine/Common/JobManager/APIHandler/AbstractAPIClient.php]: @jeawhanlee I think you forgot that

File: inc/Engine/Common/JobManager/Managers/AbstractManager.php [hunk 20]
```diff
@@ -0,0 +1,249 @@
+<?php
+
+namespace WP_Rocket\Engine\Common\JobManager\Managers;
+
+use WP_Rocket\Engine\Common\Context\ContextInterface;
+
+class AbstractManager {
```
Comment (CrochetFeve0251) [inline inc/Engine/Common/JobManager/Managers/AbstractManager.php]: As we have an interface here I am thinking that using a trait could be beneficial for later evolution as with traits we are not limited to one parent

**Review by remyperona — CHANGES_REQUESTED** [review]

File: inc/Engine/Common/JobManager/APIHandler/AbstractAPIClient.php (hunk 4 above)
Comment (vmanthos) [reply inc/Engine/Common/JobManager/APIHandler/AbstractAPIClient.php]: Why don't we do it like here?
https://github.com/wp-media/wp-rocket/blob/fb65f675bf69ba50e40918c57312e20dec0ddc87/inc/Engine/Optimization/RUCSS/AbstractAPIClient.php#L69-L71

That way, for testing we'll be able to use the constant, and we will not risk shipping the staging site URL.

File: inc/Engine/Common/Database/Queries/AbstractQuery.php (hunk 1 above)
Comment (jeawhanlee) [reply inc/Engine/Common/Database/Queries/AbstractQuery.php:169]: I'm doing this to return all the columns then use this to check which job is unique to lcp or rucss for example, row for atf will return a column with lcp and row of rucss will return a column with css.

File: inc/Engine/Common/Database/Tables/AbstractTable.php [hunk 21]
```diff
@@ -0,0 +1,140 @@
+<?php
+
+namespace WP_Rocket\Engine\Common\Database\Tables;
+
+use WP_Rocket\Dependencies\Database\Table;
+use WP_Rocket\Engine\Common\Database\Table as TableInterface;
+
+class AbstractTable extends Table implements TableInterface {
```
Comment (CrochetFeve0251) [inline inc/Engine/Common/Database/Tables/AbstractTable.php:8]: Same thing as for the manager here as we got the interface we could use a trait at that level

**Review by CrochetFeve0251 — APPROVED** [review]
As per discussion with @jeawhanlee the routing from old hooks to new hooks will be done later on.
Same for testing.

File: inc/Engine/Common/Database/Table.php [hunk 22]
```diff
@@ -0,0 +1,40 @@
+<?php
+
+namespace WP_Rocket\Engine\Common\Database;
+
+interface Table {
```
Comment (remyperona) [inline inc/Engine/Common/Database/Table.php]:
```suggestion
interface TableInterface {
```

File: inc/Engine/Common/JobManager/AbstractFactory/SaasFactory.php [hunk 23]
```diff
@@ -0,0 +1,23 @@
+<?php
+declare( strict_types=1 );
+
+namespace WP_Rocket\Engine\Common\JobManager\AbstractFactory;
+
+use WP_Rocket\Engine\Common\JobManager\Managers\ManagerInterface;
+use WP_Rocket\Engine\Common\Database\Table;
```
Comment (remyperona) [inline inc/Engine/Common/JobManager/AbstractFactory/SaasFactory.php]:
```suggestion
use WP_Rocket\Engine\Common\Database\TableInterface;
```

File: inc/Engine/Common/JobManager/AbstractFactory/SaasFactory.php [hunk 24]
```diff
@@ -0,0 +1,23 @@
...
+	 * SaaS job manager.
+	 *
+	 * @return ManagerInterface
+	 */
+	public function manager(): ManagerInterface;
+
+	/**
+	 * Job table.
+	 *
+	 * @return Table
```
Comment (remyperona) [inline inc/Engine/Common/JobManager/AbstractFactory/SaasFactory.php]:
```suggestion
	 * @return TableInterface
```

File: inc/Engine/Common/JobManager/AbstractFactory/SaasFactory.php [hunk 25]
```diff
@@ -0,0 +1,23 @@
...
+	 */
+	public function table(): Table;
```
Comment (remyperona) [inline inc/Engine/Common/JobManager/AbstractFactory/SaasFactory.php]:
```suggestion
	public function table(): TableInterface;
```

File: inc/Engine/Media/AboveTheFold/Jobs/Factory.php [hunk 26]
```diff
@@ -0,0 +1,54 @@
+<?php
+declare(strict_types=1);
+
+namespace WP_Rocket\Engine\Media\AboveTheFold\Jobs;
+
+use WP_Rocket\Engine\Common\JobManager\AbstractFactory\SaasFactory;
+use WP_Rocket\Engine\Common\JobManager\Managers\ManagerInterface;
+use WP_Rocket\Engine\Common\Database\Table;
```
Comment (remyperona) [inline inc/Engine/Media/AboveTheFold/Jobs/Factory.php]:
```suggestion
use WP_Rocket\Engine\Common\Database\TableInterface;
```

File: inc/Engine/Media/AboveTheFold/Jobs/Factory.php [hunk 27]
```diff
@@ -0,0 +1,54 @@
...
+	 * ATF Manager.
+	 *
+	 * @var ManagerInterface
+	 */
+	private $manager;
+
+	/**
+	 * ATF Table.
+	 *
+	 * @var Table
```
Comment (remyperona) [inline inc/Engine/Media/AboveTheFold/Jobs/Factory.php]:
```suggestion
	 * @var TableInterface
```

File: inc/Engine/Media/AboveTheFold/Jobs/Factory.php [hunk 28]
```diff
@@ -0,0 +1,54 @@
...
+	 */
+	private $table;
+
+	/**
+	 * Instatiate the class.
+	 *
+	 * @param ManagerInterface $manager ATF Manager.
+	 * @param Table            $table ATF Table.
```
Comment (remyperona) [inline inc/Engine/Media/AboveTheFold/Jobs/Factory.php]:
```suggestion
	 * @param TableInterface   $table ATF Table.
```

File: inc/Engine/Media/AboveTheFold/Jobs/Factory.php [hunk 29]
```diff
@@ -0,0 +1,54 @@
...
+	 */
+	public function __construct( ManagerInterface $manager, Table $table ) {
```
Comment (remyperona) [inline inc/Engine/Media/AboveTheFold/Jobs/Factory.php]:
```suggestion
	public function __construct( ManagerInterface $manager, TableInterface $table ) {
```

File: inc/Engine/Media/AboveTheFold/Jobs/Factory.php [hunk 30]
```diff
@@ -0,0 +1,54 @@
...
+	 * @return ManagerInterface
+	 */
+	public function manager(): ManagerInterface {
+		return $this->manager;
+	}
+
+	/**
+	 * ATF Table.
+	 *
+	 * @return Table
```
Comment (remyperona) [inline inc/Engine/Media/AboveTheFold/Jobs/Factory.php]:
```suggestion
	 * @return TableInterface
```

File: inc/Engine/Media/AboveTheFold/Jobs/Factory.php [hunk 31]
```diff
@@ -0,0 +1,54 @@
...
+	 */
+	public function table(): Table {
```
Comment (remyperona) [inline inc/Engine/Media/AboveTheFold/Jobs/Factory.php]:
```suggestion
	public function table(): TableInterface {
```

**Review by remyperona — CHANGES_REQUESTED** [review]

**Review by remyperona — APPROVED** [review]

**Pull request merged**